src/ensembl/src/shared/components/upload/Upload.tsx

```diff
@@ -0,0 +1,167 @@
+  let dragCounter = 0;
+
+  const handleDragIn = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    dragCounter++;
+    if (get(e, 'dataTransfer.items.length', 0) > 0) {
+      setDrag(true);
+    }
+  };
+
+  const handleDragOut = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    dragCounter--;
+    if (dragCounter === 0) {
+      setDrag(false);
+    }
+  };
+
+  const handleDrop = (e: React.DragEvent) => {
+    e.preventDefault();
+    e.stopPropagation();
+    setDrag(false);
+
+    if (
+      e.dataTransfer &&
+      e.dataTransfer.files &&
+      e.dataTransfer.files.length > 0
+    ) {
+      if (props.callbackWithFiles) {
+        if (!props.allowMultiple) {
+          props.onChange(e.dataTransfer.files[0]);
+          return;
+        }
+        props.onChange(e.dataTransfer.files);
+        return;
+      }
+
+      const { files } = e.dataTransfer;
+
+      forEach(files, (file) => {
+        const fileReader = new FileReader();
+        fileReaders.push(fileReader);
+        totalPendingFilesToRead++;
+        fileReader.onloadend = handleFileRead;
+        fileReader.readAsText(file);
+      });
+
+      e.dataTransfer.clearData();
+      dragCounter = 0;
+    }
+  };
+
+  const handleFileRead = () => {
+    totalPendingFilesToRead--;
+    // Do not return the content until all files are read
+    if (totalPendingFilesToRead > 0 || props.callbackWithFiles) {
+      return;
+    }
+
+    const content: string[] = fileReaders.map(
+      (fileReader) => fileReader.result as string
+    );
+
+    props.onChange(content.join('\n'));
+  };
+
+  const handleFileChosen = (e: React.ChangeEvent<HTMLInputElement>) => {
+    const files: FileList | null = get(e, 'target.files', null);
+
+    if (!files) {
+      return;
+    }
+    if (props.callbackWithFiles) {
+      // Just consider the first file if allowMultiple is true
+      if (!props.allowMultiple) {
+        props.onChange(files[0]);
+        return;
+      }
+      props.onChange(files);
+      return;
+    }
+
+    forEach(files, (file) => {
+      const fileReader = new FileReader();
+      fileReaders.push(fileReader);
+      totalPendingFilesToRead++;
+      fileReader.onloadend = handleFileRead;
+      fileReader.readAsText(file);
+    });
+  };
+
+  const className = classNames(styles.defaultUpload, props.className);
+
+  return (
+    <span
+      className={className}
+      onDragEnter={handleDragIn}
+      onDragLeave={handleDragOut}
+      onDrop={handleDrop}
+    >
+      {drag && <Overlay />}
+      <input
+        type="file"
+        id={props.id}
+        name={props.name}
+        className={styles.fileInput}
+        onChange={(e) => handleFileChosen(e)}
+        multiple={props.allowMultiple}
+      />
+      <label htmlFor={props.id}>{props.label}</label>
+    </span>
+  );
+};
+
+Upload.defaultProps = {
+  callbackWithFiles: false,
+  allowMultiple: true,
+  id: 'file',
```

Comment:
erm, why would we have default id? that means we can have multiple inputs with the same id on the page, and this goes against html spec

Reply:
I need a default id that can be used in the `htmlFor` attribute of the `label` tag. I guess I will have to generate a unique ID for every Upload component?

Reply:
You are wrapping your input in a label. I believe you don't even need the htmlFor in this case — the click on a label will toggle the input by itself.

Reply:
Yes, that's right. I've removed it.